akka-http/src/main/scala/akka/http/scaladsl/server/directives/JwtDirectives.scala

```diff
@@ -0,0 +1,98 @@
+package akka.http.scaladsl.server.directives
+
+import akka.http.impl.util.enhanceString_
+import akka.http.scaladsl.common.{ NameDefaultReceptacle, NameOptionReceptacle, NameReceptacle, RequiredValueReceptacle }
+import akka.http.scaladsl.model.{ AttributeKey, HttpRequest }
+import akka.http.scaladsl.server.{ Directive, Directive0, Directive1, InvalidRequiredValueForQueryParamRejection, MalformedQueryParamRejection, MalformedRequestContentRejection, MissingQueryParamRejection, RequestContext }
+import akka.http.scaladsl.server.Directives.{ Authenticator, AuthenticatorPF, authenticateOAuth2, authenticateOAuth2Async, authenticateOAuth2PF, headerValueByName }
+import akka.http.scaladsl.server.directives.BasicDirectives.{ extractRequestContext, pass, provide }
+import akka.http.scaladsl.server.directives.FutureDirectives.onComplete
+import akka.http.scaladsl.server.directives.RouteDirectives.reject
+import akka.http.scaladsl.server.util.{ JwtSprayJson, JwtSupport }
+import akka.http.scaladsl.unmarshalling.Unmarshaller
+import akka.http.scaladsl.util.FastFuture
+import com.typesafe.config.Config
+import pdi.jwt.JwtClaim
+import spray.json.{ JsBoolean, JsNumber, JsObject, JsString, JsValue }
+
+import scala.concurrent.duration.DurationInt
+import scala.concurrent.{ Await, Future }
+import scala.reflect.ClassTag
+import scala.util.{ Failure, Success }
+
+trait JwtDirectives {
+
+  import BasicDirectives._
+  import RouteDirectives._
+
+  def jwt(): Directive1[JwtClaims] = {
+    extractActorSystem.flatMap { system =>
+      jwt(system.settings.config)
+    }
+  }
+
+  private def jwt(settings: Config): Directive1[JwtClaims] = {
+    // FIXME: Should we use this or just manually export the header value? Is there a clever way to extract a realm that makes sense? path?
+    authenticateOAuth2("realm", bearerTokenAuthenticator(settings)).flatMap { claims =>
+      extractRequestContext.flatMap { ctx =>
+        provide(new JwtClaims(ctx, claims))
+      }
+    }
+  }
+
+  private def bearerTokenAuthenticator(settings: Config): Authenticator[JsObject] = {
+    case p @ Credentials.Provided(token) =>
+      val jwtSupport = JwtSupport.fromConfig(settings)
+
+      jwtSupport.validate(token) match {
+        case Right(claims) => Some(claims)
+        case Left(ex)      => None // FIXME: how to propagate this?
+      }
+  }
+
+}
+
+// JwtClaims provides utilities to easily assert and extract claims from the JWT token
+class JwtClaims(ctx: RequestContext, claims: JsObject) {
+
+  import akka.http.scaladsl.unmarshalling.{ FromStringUnmarshaller => FSU }
+  import ctx.executionContext
+  import ctx.materializer
+
+  def hasClaim(name: String): Boolean = claims.fields.contains(name)
+
+  def get[T](name: String)(implicit fsu: FSU[T]): Future[T] = getClaim[T](name)
```

Comment:
I think we should skip the integration with `FromStringUnmarshaller` here, having to deal with `Future` is completely superficial, just having a few concrete methods for `getInt`, `getString`, `getBoolean` that all returns `Option` is enough and easily discoverable, handling default values is already built into `Option`/`Optional` so no need for an additional method for that.

Maybe name them `intClaim` for Scala and `getIntClaim` for Java - that way there could be one interface per language but a common implementation.

Reply:
Thanks. removed `FromStringUnmarshaller` now, still lacking a common interface. Will add when implementing the java api.